### PR TITLE
Move MappedFieldType.similarity() to TextSearchInfo

### PR DESCRIPTION
**Pull request by romseygeek**
Similarities only apply to a few text-based field types, but are currently set directly on
the base MappedFieldType class.  This commit moves similarity information into
TextSearchInfo, and removes any mentions of it from MappedFieldType or FieldMapper.

It was previously possible to include a `similarity` parameter on a number of field types 
that would then ignore this information.  To make it obvious that this has no effect, setting
this parameter on non-text field types now issues a deprecation warning.

### CODE REVIEW

File: modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java [hunk 1]
```diff
@@ -113,30 +116,39 @@
...
                     builder.maxShingleSize(nodeIntegerValue(fieldNode));
                     iterator.remove();
+                } else if (fieldName.equals("similarity")) {
+                    SimilarityProvider similarityProvider = TypeParsers.resolveSimilarity(parserContext, fieldName, fieldNode.toString());
+                    builder.similarity(similarityProvider);
+                    iterator.remove();
                 }
                 // TODO should we allow to configure the prefix field
             }
+            parseTextField(builder, name, node, parserContext);
```
Comment (romseygeek) [inline modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java:134]: The shared `parseTextField()` method does a deprecation check for `similarity` now, so we need to move it to after we've processed the `similarity` field.  To ensure that mapper-specific fields still get checked for `null`, a new `TypeParsers.checkNull()` method is called for each entry.

File: server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java [hunk 2]
```diff
@@ -310,8 +317,13 @@ public TextFieldMapper build(BuilderContext context) {
                 } else if (propName.equals("index_phrases")) {
                     builder.indexPhrases(XContentMapValues.nodeBooleanValue(propNode, "index_phrases"));
                     iterator.remove();
+                } else if (propName.equals("similarity")) {
+                    SimilarityProvider similarityProvider = TypeParsers.resolveSimilarity(parserContext, fieldName, propNode.toString());
+                    builder.similarity(similarityProvider);
+                    iterator.remove();
                 }
             }
+            parseTextField(builder, fieldName, node, parserContext);
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java:326]: Same process here as in SearchAsYouTypeFieldMapper

File: test/framework/src/main/java/org/elasticsearch/index/mapper/FieldMapperTestCase.java [hunk 3]
```diff
@@ -106,10 +103,6 @@ private Modifier booleanModifier(String name, boolean updateable, BiConsumer<T,
             a.searchQuoteAnalyzer(new NamedAnalyzer("standard", AnalyzerScope.INDEX, new StandardAnalyzer()));
             a.searchQuoteAnalyzer(new NamedAnalyzer("whitespace", AnalyzerScope.INDEX, new WhitespaceAnalyzer()));
         }),
-        new Modifier("similarity", false, (a, b) -> {
-            a.similarity(new SimilarityProvider("BM25", new BM25Similarity()));
-            b.similarity(new SimilarityProvider("boolean", new BooleanSimilarity()));
-        }),
```
Comment (romseygeek) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/FieldMapperTestCase.java:112]: This has been moved directly into the applicable FieldMappers

**Review by nik9000 — APPROVED** [review]
LGTM

**Pull request merged**